### ISSUE TITLE
Provider automatically associate already existing resources with multiple separate terraform states with same resource names

### ISSUE THREAD

**Issue by Cyanopus** [terraform-providers/terraform-provider-azurerm#7972]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform 0.12.28
Azure Provider: 2.15.0

### Affected Resource(s)

* any
Tested with:
* azurerm_resource_group ( very critical, we've removed the resource_group from our modules as this bug would wipe part of our business - basically any setup that hooked the resource group to them when run a simple destroy for test would wipe the whole resource group with the rest of the setups)
* azurerm_virtual_network
* azurerm_application_gateway
* azurerm_kubernetes_cluster

### Terraform Configuration Files

Just regular usage but following the same resource name and resource group.
/subscriptions/example1/resourceGroups/example1/providers/Microsoft.ContainerService/managedClusters/example1/
/subscriptions/example1/resourceGroups/example1/providers/Microsoft.ContainerService/managedClusters/example1/

This situation can happen if you have big setup in one resource group with 2 clusters for instance with the same name( due to false input for instance) but in different states as the two setups are in same resourceGroup but provisioned by 2 separate terraform codes and backends.

### Debug Output

 
Initializing the backend...
 
Successfully configured the backend "s3"! Terraform will automatically
use this backend unless the backend configuration changes.
 
Initializing provider plugins...
- Checking for available provider plugins...
- Downloading plugin for provider "azurerm" (hashicorp/azurerm) 2.15.0...
 
Terraform has been successfully initialized!
Switched to workspace "example1".
data.terraform_remote_state.virtual_network: Refreshing state...
module.azure_aks_cluster.azurerm_log_analytics_workspace.aks[0]: Creating...
module.azure_aks_cluster.azurerm_log_analytics_workspace.aks[0]: Still creating... [10s elapsed]
module.azure_aks_cluster.azurerm_log_analytics_workspace.aks[0]: Still creating... [20s elapsed]
module.azure_aks_cluster.azurerm_log_analytics_workspace.aks[0]: Still creating... [30s elapsed]
module.azure_aks_cluster.azurerm_log_analytics_workspace.aks[0]: Still creating... [40s elapsed]
module.azure_aks_cluster.azurerm_log_analytics_workspace.aks[0]: Still creating... [50s elapsed]
module.azure_aks_cluster.azurerm_log_analytics_workspace.aks[0]: Still creating... [1m0s elapsed]
module.azure_aks_cluster.azurerm_log_analytics_workspace.aks[0]: Creation complete after 1m5s [id=/subscriptions/example1/resourcegroups/example1/providers/microsoft.operationalinsights/workspaces/example1-law]
module.azure_aks_cluster.azurerm_log_analytics_solution.aks[0]: Creating...
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Creating...
module.azure_aks_cluster.azurerm_log_analytics_solution.aks[0]: Creation complete after 3s [id=/subscriptions/example1/resourcegroups/example1/providers/Microsoft.OperationsManagement/solutions/ContainerInsights(example1-law)]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [10s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [20s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [30s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [40s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [50s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [1m0s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [1m10s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [1m20s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [1m30s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [1m40s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [1m50s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [2m0s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [2m10s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [2m20s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [2m30s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [2m40s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [2m50s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [3m0s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [3m10s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [3m20s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [3m30s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [3m40s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [3m50s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [4m0s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Still creating... [4m10s elapsed]
module.azure_aks_cluster.azurerm_kubernetes_cluster.k8s: Creation complete after 4m14s [id=/subscriptions/example1/resourcegroups/example1/providers/Microsoft.ContainerService/managedClusters/example1]

Cluster is provisioned in "lightning fast" fashion for 4 minutes, which is far less than the baseline time for the service.
Checking then in the console you see that no cluster was provided but it is the one that was already there but was just returned by the API.
### Panic Output

There is no Panic output and that's the problem.

### Expected Behavior

When a resource with such id is already existing, the provider must error out and point that fact so the user can change the name to avoid 2 different terraform setups and states to associate with the same resource.

### Actual Behavior

Terraform runs for a fast time and associates already existing resources from 1 state to another which causes race condition on the resource in two different to terraform setups. Potentially one setup can destroy the resource of another setup just because this was allowed and associated by the API and the provider.

### Steps to Reproduce

1. Create terraform code that creates 2 resources with static names, same resource group, same subscription. Copy it twice in different folders.
2. Add two different backend configurations on the 2 folders ( with remote state you must specify different keys for the path)
3. Apply both one after another.
4. Monitor the second execution and note that your resources are taking considerably less time.
5. Check the Azure portal. You will see one set of resources
6. Check the terraform states and see that they are associated to the same resources in the console.

**Comment by njuCZ** [comment]
@Cyanopus is there any business scenario that you need to run the same TF scripts in different workspace simultaneously? in fact, we don't recommend to do like this. 

Currently, before invoking create function of every resources, we would check whether it exists. In your case, both process are running simultaneously, so both of them find the resource doesn't exist, and try to create it. In this way, it really depends on the backend service implementation. If the backend service api is Idempotent，it's possible that they will not report error, which lead to the result that two different state links to the same resources.

Anyway, we don't recommend to do like this

**Comment by Cyanopus** [comment]
@njuCZ We are not creating the resources simultaneously most of the time one of the resource was created some time ago and just an automated test or provisioning happens and depending on the input params match the old one. They are completely independent the only share is the resource group. And we also run separate workspaces to isolate the resource one another. We face this mostly in our development endeavors as we use one resource group for dynamic provisioning and destroy. But also we faced this with a project that was having two env dev and prod in the same resource group and we had to cope with it and remove the resource group at all from the network module as deleting dev would result in wasting prod as well. The problem is that you cannot stop someone from creating a resource group with the same name in a different terrafrom state, but backend/api/provider must return information that such resource group already exist or assure unique ID in the backend, not just assign it to whoever asks for resource group with such name. We are working in many cloud providers and this is the first cloud that actually does that. It is not possible to always assure that you are not going to hit someone else work. Is it recommended to go to the console and verify if any of those resources do not pre-exist? I find this odd.

About the business case, yes we are running multiple platforms using dynamic terraform code and this can technically happen and happens.
